native/dtrevc3.go

```diff
@@ -0,0 +1,862 @@
+// VL will contain:
+//  if howmny == lapack.AllEigVec,      the matrix Y of left eigenvectors of T,
+//  if howmny == lapack.AllEigVecMulQ,  the matrix Q*Y,
+//  if howmny == lapack.SelectedEigVec, the left eigenvectors of T specified by
+//                                      selected, stored consecutively in
+//                                      the columns of VL, in the same order
+//                                      as their eigenvalues.
+// VL is not referenced if side == lapack.RightEigVec.
+//
+// On return, if side is lapack.RightEigVec or lapack.RightLeftEigVec,
+// VR will contain:
+//  if howmny == lapack.AllEigVec,      the matrix X of right eigenvectors of T,
+//  if howmny == lapack.AllEigVecMulQ,  the matrix Q*X,
+//  if howmny == lapack.SelectedEigVec, the left eigenvectors of T specified by
+//                                      selected, stored consecutively in
+//                                      the columns of VR, in the same order
+//                                      as their eigenvalues.
+// VR is not referenced if side == lapack.LeftEigVec.
+//
+// Complex eigenvectors corresponding to a complex eigenvalue are stored in VL
+// and VR in two consecutive columns, the first holding the real part, and the
+// second the imaginary part.
+//
+// Each eigenvector will be normalized so that the element of largest magnitude
+// has magnitude 1. Here the magnitude of a complex number (x,y) is taken to be
+// |x| + |y|.
+//
+// work must have length at least lwork and lwork must be at least max(1,3*n),
+// otherwise Dtrevc3 will panic. For optimum performance, lwork should be at
+// least n+2*n*nb, where nb is the optimal blocksize.
+//
+// If lwork == -1, instead of performing Dtrevc3, the function only estimates
+// the optimal workspace size based on n and stores it into work[0].
+//
+// Dtrevc3 returns the number of columns in VL and/or VR actually used to store
+// the eigenvectors.
+//
+// Dtrevc3 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dtrevc3(side lapack.EigVecSide, howmny lapack.HowMany, selected []bool, n int, t []float64, ldt int, vl []float64, ldvl int, vr []float64, ldvr int, mm int, work []float64, lwork int) (m int) {
+	switch side {
+	default:
+		panic(badSide)
+	case lapack.RightEigVec, lapack.LeftEigVec, lapack.RightLeftEigVec:
+	}
+	switch howmny {
+	default:
+		panic(badHowMany)
+	case lapack.AllEigVec, lapack.AllEigVecMulQ, lapack.SelectedEigVec:
+	}
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork < max(1, 3*n) && lwork != -1:
+		panic(badWork)
+	}
+	if lwork != -1 {
+		if howmny == lapack.SelectedEigVec {
+			if len(selected) != n {
+				panic("lapack: bad selected length")
+			}
+			// Set m to the number of columns required to store the
+			// selected eigenvectors, and standardize the slice
+			// selected.
+			for j := 0; j < n; {
```

Comment:
Yeah, I've written this real/complex eigenvalue distingushing loop so many times by now that it's become a second nature. I've tried to follow the fortran original more closely and keep the pair variable but it looks worse than what I have here. I've added two comments and removed the `continue` statement so that now the two branches are directly comparable. Let me know what you think.